File: sdks/python/examples/sidecar_charge.py
#!/usr/bin/env python3
"""patala from Python over the sidecar — a separate process, no FFI at all.

    python3 sdks/python/examples/sidecar_charge.py

This spawns `patala-sidecar` on a free loopback port with a freshly generated
token, drives a full quote -> charge -> verify round trip against it with
nothing but the standard library, and shuts it down again. Nothing is left
running when it exits.

Everything runs against `MockRail` — deterministic, offline, no credentials.

Why you might pick this over ../../patala-py, even though patala's in-process
cost is unusually low (see README.md):

  - **Key isolation.** A non-custodial rail's signing key lives in whichever
    process calls charge. One sidecar means one process holds it, instead of
    every Python worker in the fleet.
  - No Rust toolchain, no wheel, no platform matrix on the calling side — just
    a binary and an HTTP client.
  - The exact same JSON the C ABI speaks, so a call site moves between the two
    modes as a transport change, not a rewrite.

Build the server first, from the workspace root:

    cargo build -p patala-sidecar
"""

from __future__ import annotations

import json
import os
import secrets
import socket
import subprocess
import sys
import time
import urllib.error
import urllib.request
from pathlib import Path

REPO = Path(__file__).resolve().parents[3]
CHECKS = [0]


def check(condition: bool, message: str) -> None:
    CHECKS[0] += 1
    if not condition:
        sys.exit(f"FAILED: {message}")
    print(f"  ok  {message}")


def resolve_binary() -> str:
    """PATALA_SIDECAR_BIN, then a workspace build, then PATH."""
    env = os.environ.get("PATALA_SIDECAR_BIN")
    if env:
        return env
    for profile in ("debug", "release"):
        candidate = REPO / "target" / profile / "patala-sidecar"
        if candidate.is_file():
            return str(candidate)
    return "patala-sidecar"


def free_port() -> int:
    with socket.socket() as sock:
        sock.bind(("127.0.0.1", 0))
        return sock.getsockname()[1]


class Sidecar:
    """Spawn `patala-sidecar`, wait for /healthz, terminate it on the way out."""

    def __init__(self) -> None:
        self.token = secrets.token_hex(32)
        self.port = free_port()
        self.base = f"http://127.0.0.1:{self.port}"
        self.binary = resolve_binary()
        env = dict(os.environ, PATALA_SIDECAR_TOKEN=self.token,
                   PATALA_SIDECAR_PORT=str(self.port))
        try:
            self.proc = subprocess.Popen(  # noqa: S603
                [self.binary], env=env,
                stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL,
            )
        except FileNotFoundError:
            sys.exit(
                f"could not run {self.binary!r}.\n"
                "Build it first: cargo build -p patala-sidecar"
            )

    def __enter__(self) -> "Sidecar":
        deadline = time.monotonic() + 10
        while time.monotonic() < deadline:
            if self.proc.poll() is not None:
                sys.exit(f"patala-sidecar exited with {self.proc.returncode} before answering")
            try:
                with urllib.request.urlopen(f"{self.base}/healthz", timeout=0.5) as resp:
                    if resp.read() == b"ok":
                        return self
            except (urllib.error.URLError, OSError):
                time.sleep(0.05)
        self.__exit__(None, None, None)
        sys.exit("patala-sidecar never became healthy")

    def __exit__(self, *_exc: object) -> None:
        self.proc.terminate()
        try:
            self.proc.wait(timeout=5)
        except subprocess.TimeoutExpired:
            self.proc.kill()
            self.proc.wait()

    def request(self, method: str, path: str, body: object = None,
                token: str | None = ...) -> tuple[int, object]:  # type: ignore[assignment]
        """One HTTP call. Returns (status, decoded-body-or-text).

        A 4xx/5xx is returned rather than raised: on this API a non-2xx is an
        answer with a body worth reading, not a transport failure.
        """
        data = None if body is None else json.dumps(body).encode()
        req = urllib.request.Request(f"{self.base}{path}", data=data, method=method)
        if token is ...:
            token = self.token
        if token is not None:
            req.add_header("Authorization", f"Bearer {token}")
        if data is not None:
            req.add_header("Content-Type", "application/json")
        try:
            with urllib.request.urlopen(req, timeout=10) as resp:
                return resp.status, _decode(resp.read())
        except urllib.error.HTTPError as exc:
            return exc.code, _decode(exc.read())


def _decode(raw: bytes) -> object:
    try:
        return json.loads(raw)
    except json.JSONDecodeError:
        return raw.decode(errors="replace")


def main() -> int:
    with Sidecar() as sc:
        print(f"binary:   {sc.binary}")
        print(f"listening on {sc.base} (loopback only — the bind address is not configurable)")
        print(f"python:   {sys.version.split()[0]} on {sys.platform}\n")

        print("capabilities")
        status, caps = sc.request("GET", "/v1/rails/mock")
        check(status == 200, f"GET /v1/rails/mock -> {status}")
        check(
            caps["class"] == "NonCustodialFinal",
            f"class is {caps['class']!r} — decide the whole UX off this, not off a provider name",
        )
        check(caps["holds_funds"] is False, "holds_funds is false")

        print("\npre-flight: validate-destination, before any money moves")
        status, verdict = sc.request(
            "POST", "/v1/rails/mock/validate-destination",
            {"destination": "mock:wallet:alice"},
        )
        check(status == 200 and verdict["status"] == "StructurallyValid",
              f"a well-formed address -> 200 {verdict['status']!r}")
        check(verdict["is_refusal"] is False,
              "is_refusal is false — read the body, not just the status code")
        check(verdict["human_must_confirm"] is True,
              "human_must_confirm is true even on StructurallyValid")
        status, refused = sc.request(
            "POST", "/v1/rails/mock/validate-destination", {"destination": ""},
        )
        check(status == 200 and refused["status"] == "Malformed" and refused["is_refusal"],
              "an empty destination is a well-formed REQUEST -> 200 with a Malformed refusal")

        print("\nquote -> charge -> verify")
        pay = {"amount_minor": 1250, "currency": "USDC",
               "destination": "mock:wallet:alice", "reference": "order-1"}
        status, quote = sc.request("POST", "/v1/rails/mock/quote", pay)
        check(status == 200 and quote["total_minor"] == 1250,
              f"total_minor == {quote['total_minor']}")
        check(isinstance(quote["total_minor"], int),
              "the JSON number decodes to an int — minor units, never a float")

        status, receipt = sc.request("POST", "/v1/rails/mock/charge", pay)
        check(status == 200 and receipt["amount_minor"] == 1250,
              f"charge -> receipt for {receipt['amount_minor']} {receipt['currency']}")

        status, verdict = sc.request("POST", "/v1/rails/mock/verify", receipt)
        check(status == 200 and verdict == {"valid": True},
              "the genuine receipt verifies {'valid': true}")

        tampered = dict(receipt, amount_minor=receipt["amount_minor"] + 1)
        status, verdict = sc.request("POST", "/v1/rails/mock/verify", tampered)
        check(status == 200 and verdict == {"valid": False},
              "a tampered receipt is 200 {'valid': false} — fail-closed, and NOT an HTTP error")

        print("\nthe error surface, so you can tell these four apart")
        status, body = sc.request("POST", "/v1/rails/mock/charge", dict(pay, currency="EUR"))
        check(status == 400, f"an unsupported currency -> {status} {body['kind']!r}")
        status, body = sc.request("GET", "/v1/rails/nope")
        check(status == 404, f"an unknown rail_id -> {status} {body['kind']!r}")
        status, body = sc.request("POST", "/v1/rails/mock/webhook", {})
        check(status == 501,
              f"the mock has no push delivery -> {status} {body['kind']!r}, never an invented event")
        status, body = sc.request("GET", "/v1/rails/mock", token=None)
        check(status == 401, f"no Authorization header -> {status} on a READ-ONLY route too")

    print(f"\nsidecar terminated; nothing left running")
    print(f"\nALL {CHECKS[0]} PYTHON SIDECAR ASSERTIONS PASSED")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: sdks/elixir/examples/sidecar_charge.exs
# patala from Elixir over the sidecar — a separate OS process, no NIF at all.
#
#     cd sdks/elixir && mix run examples/sidecar_charge.exs
#
# This spawns patala-sidecar on a free loopback port with a freshly generated
# token, drives a full quote -> charge -> verify against MockRail over
# :gen_tcp, and terminates it. Nothing is left running, and nothing outside the
# standard library is used — not even :inets.
#
# THIS IS THE RECOMMENDED DEFAULT FOR ELIXIR, and the last section shows why in
# the only terms that matter on the BEAM: a call you can abandon, a failure you
# can supervise, and a fault that cannot take the VM.
#
# Build the server first, from the workspace root:
#
#     cargo build -p patala-sidecar

defmodule SidecarExample do
  alias Patala.Sidecar

  @pay %{
    amount_minor: 1250,
    currency: "USDC",
    destination: "mock:wallet:alice",
    reference: "order-1"
  }

  def run do
    Process.put(:checks, 0)

    IO.puts("elixir #{System.version()} / OTP #{:erlang.system_info(:otp_release)}")

    {:ok, sc} = Sidecar.spawn()

    try do
      IO.puts("binary:  #{sc.binary}")
      IO.puts("listening on #{sc.base_url} (loopback only — the bind address is not configurable)")
      IO.puts("os pid:  #{sc.os_pid}   <- a real OS process, which is the entire point\n")

      IO.puts("capabilities")
      {:ok, caps} = Sidecar.capabilities(sc, "mock")

      check(
        caps["class"] == "NonCustodialFinal",
        "class is #{inspect(caps["class"])} — decide the whole UX off this, not off a provider name"
      )

      check(caps["holds_funds"] == false, "holds_funds is false")

      IO.puts("\npre-flight: validate-destination, before any money moves")
      {:ok, verdict} = Sidecar.validate_destination(sc, "mock", "mock:wallet:alice")

      check(
        verdict["status"] == "StructurallyValid",
        "a well-formed address -> 200 #{inspect(verdict["status"])}"
      )

      check(
        verdict["is_refusal"] == false,
        "is_refusal is false — read the body, not just the code"
      )

      check(
        verdict["human_must_confirm"] == true,
        "human_must_confirm is true even on StructurallyValid"
      )

      {:ok, 200, refused} =
        Sidecar.try(sc, "POST", "/v1/rails/mock/validate-destination", %{destination: ""})

      check(
        refused["status"] == "Malformed" and refused["is_refusal"],
        "an empty destination is a well-formed REQUEST -> 200 with a Malformed refusal"
      )

      IO.puts("\nquote -> charge -> verify")
      {:ok, quote} = Sidecar.quote(sc, "mock", @pay)

      check(
        quote["total_minor"] == 1250 and is_integer(quote["total_minor"]),
        "total_minor == #{quote["total_minor"]} and decodes as an integer — never a float"
      )

      {:ok, receipt} = Sidecar.charge(sc, "mock", @pay)

      check(
        receipt["amount_minor"] == 1250,
        "charge -> receipt for #{receipt["amount_minor"]} #{receipt["currency"]}"
      )

      check(
        Sidecar.verify(sc, "mock", receipt) == {:ok, %{"valid" => true}},
        "the genuine receipt verifies {\"valid\": true}"
      )

      tampered = Map.update!(receipt, "amount_minor", &(&1 + 1))
      {:ok, 200, body} = Sidecar.try(sc, "POST", "/v1/rails/mock/verify", tampered)

      check(
        body == %{"valid" => false},
        "a tampered receipt is 200 {\"valid\": false} — fail-closed, and NOT an HTTP error"
      )

      IO.puts("\nthe error surface, so you can tell these four apart")

      {:ok, 400, body} = Sidecar.try(sc, "POST", "/v1/rails/mock/charge", %{@pay | currency: "EUR"})
      check(true, "an unsupported currency -> 400 #{inspect(body["kind"])}")

      {:ok, 404, body} = Sidecar.try(sc, "GET", "/v1/rails/nope")
      check(true, "an unknown rail_id -> 404 #{inspect(body["kind"])}")

      {:error, {501, kind, _}} = Sidecar.webhook(sc, "mock", "{}")

      check(
        kind == "unsupported",
        "the mock has no push delivery -> 501 #{inspect(kind)}, never an invented event"
      )

      {:ok, 401, _} = Sidecar.try(sc, "GET", "/v1/rails/mock", nil, authed: false)
      check(true, "no Authorization header -> 401 on a READ-ONLY route too")

      beam_properties(sc)
    after
      Sidecar.stop(sc)
    end

    IO.puts("\nsidecar terminated; nothing left running")
    IO.puts("\nALL #{Process.get(:checks)} ELIXIR SIDECAR ASSERTIONS PASSED")
  end

  # The part a NIF cannot match, demonstrated rather than argued.
  defp beam_properties(sc) do
    IO.puts("\nwhat the process boundary buys, measured")

    # 1. An in-flight call can be abandoned. Task.shutdown actually stops the
    #    work here, because the work is a socket read in a BEAM process — a NIF
    #    on a dirty scheduler would keep going after the timeout returned.
    task = Task.async(fn -> Sidecar.charge(sc, "mock", @pay) end)

    case Task.yield(task, 5_000) || Task.shutdown(task, :brutal_kill) do
      {:ok, {:ok, _receipt}} ->
        check(true, "a charge inside a Task returns normally")

      other ->
        check(false, "unexpected task result: #{inspect(other)}")
    end

    slow = Task.async(fn -> Sidecar.charge(sc, "mock", @pay) end)
    Task.shutdown(slow, :brutal_kill)

    check(
      Process.alive?(slow.pid) == false,
      "Task.shutdown/:brutal_kill really stops it — there is a PROCESS to kill, which is the whole difference"
    )

    # 2. A supervisor can restart the caller, because the caller is a process.
    {:ok, supervisor} = Task.Supervisor.start_link()

    task =
      Task.Supervisor.async_nolink(supervisor, fn ->
        Sidecar.capabilities(sc, "definitely-not-a-rail")
      end)

    check(
      match?({:ok, {:error, {404, "unknown_rail", _}}}, Task.yield(task, 5_000)),
      "a failure comes back as data to a supervised process, not as a VM-wide event"
    )

    Supervisor.stop(supervisor)

    # 3. The sidecar is a separate OS process; a fault in it cannot take this VM.
    check(
      is_integer(sc.os_pid) and sc.os_pid != System.pid() |> String.to_integer(),
      "the rail runs in OS pid #{sc.os_pid}, this VM is #{System.pid()} — a segfault there is not a segfault here"
    )
  end

  defp check(true, message) do
    Process.put(:checks, Process.get(:checks) + 1)
    IO.puts("  ok  #{message}")
  end

  defp check(_false, message) do
    IO.puts(:stderr, "FAILED: #{message}")
    System.halt(1)
  end
end

SidecarExample.run()
